wire-library/wire-compiler/src/main/java/com/squareup/wire/Manifest.kt

```diff
@@ -0,0 +1,310 @@
+  )
+
+  companion object {
+    @JvmStatic
+    val NONE = Manifest(
+        modules = mapOf("./" to Module())
+    )
+
+    private val serializer = MapSerializer(String.serializer(), Module.serializer())
+
+    fun fromYaml(string: String): Manifest {
+      return Manifest(Yaml.default.parse(serializer, string))
+    }
+  }
+}
+
+/** Module name to partition info. The iteration order of this map is the generation order. */
+class PartitionedSchema(
+  modules: Map<String, Partition>,
+  val warnings: List<String>,
+  val errors: List<String>
+) : Map<String, Partition> by modules {
+  class Partition(
+    val schema: Schema,
+    val types: Set<ProtoType>,
+    val upstreamTypes: Map<ProtoType, String>
+  )
+}
+
+private const val debug = false
+
+fun Schema.partition(manifest: Manifest): PartitionedSchema {
+  val modules = manifest.modules
+
+  if (debug) {
+    println("Modules: ${modules.keys}")
+    println("Roots: ${manifest.roots}")
+    println("Order: ${manifest.order}")
+  }
+
+  val disjointGraphs = manifest.roots
+      .map { root ->
+        val reachableNames = mutableSetOf<String>()
+        val visitQueue = ArrayDeque<String>().apply { add(root) }
+        while (visitQueue.isNotEmpty()) {
+          val visitName = visitQueue.removeFirst()
+          reachableNames += visitName
+
+          val upstreamModules = modules.getValue(visitName).dependencies
+          val downstreamModules = modules.filterValues { visitName in it.dependencies }.keys
+          for (dependency in upstreamModules + downstreamModules) {
+            if (dependency !in reachableNames && dependency !in visitQueue) {
+              visitQueue += dependency
+            }
+          }
+        }
+        // Sort the reachable names by their module order.
+        manifest.order.filter { it in reachableNames }
+      }
+      .toSet()
+  if (debug) {
+    println("Disjoint Graphs:")
+    for (disjointGraph in disjointGraphs) {
+      println(" - $disjointGraph")
+    }
+  }
+
+  val errors = mutableListOf<String>()
+  val partitions = mutableMapOf<String, Partition>()
+  for (moduleName in manifest.order) {
+    val module = modules.getValue(moduleName)
+
+    val transitiveDependencySet = mutableSetOf<String>().apply {
+      val queue = ArrayDeque<String>().apply { add(moduleName) }
+      while (queue.isNotEmpty()) {
+        val dependency = modules.getValue(queue.removeFirst())
+        addAll(dependency.dependencies)
+        queue.addAll(dependency.dependencies)
+      }
+    } as Set<String> // TODO use buildSet
+
+    val pruningRules = PruningRules.Builder()
+        .apply {
+          for (dependencyName in transitiveDependencySet) {
+            val dependency = modules.getValue(dependencyName)
+            addRoot(dependency.roots)
+            prune(dependency.prunes)
+          }
+          addRoot(module.roots)
+          // TODO(jw): what do we do about prunes that apply to types in an upstream module? fail?
```

Comment:
you should be able to detect this by seeing if any type is a subset of itself in a different module

Reply:
The intent is to get rid of this inheritance, actually. May be worth omitting it from this PR and inverting the behavior where prunes are local but we generate unnecessary types.

This manifests as the `@Ignore` in the partition test moving from the current method to the other one.

Reply:
Remove this for now. It would have gone away anyway.